src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -349,5 +411,152 @@ internal static bool TryParseSeverity(string value, out ReportDiagnostic severit
+                }
+
+                MergeSection(config.PathToFile, config.GlobalSection);
+                foreach (var section in config.NamedSections)
+                {
+                    MergeSection(config.PathToFile, section);
+                }
+            }
+
+            internal GlobalAnalyzerConfig? Build(DiagnosticBag diagnostics)
+            {
+                if (_values is null || _duplicates is null)
+                {
+                    return null;
+                }
+
+                // issue diagnostics for any duplicate keys
+                foreach ((var section, var keys) in _duplicates)
+                {
+                    bool isGlobalSection = string.IsNullOrWhiteSpace(section);
+                    string sectionName = isGlobalSection ? "Global Section" : section;
+                    foreach ((var keyName, var configPaths) in keys)
+                    {
+                        diagnostics.Add(Diagnostic.Create(
+                             MultipleGlobalAnalyzerKeysDescriptor,
+                             Location.None,
+                             keyName,
+                             sectionName,
+                             string.Join(", ", configPaths)));
+                    }
+                }
+                _duplicates.Free();
+
+                // gather the global and named sections
+                Section globalSection = getSection(string.Empty);
+                _values.Remove(string.Empty);
+
+                ArrayBuilder<Section> namedSectionBuilder = new ArrayBuilder<Section>(_values.Count);
+                foreach (var sectionName in _values.Keys)
+                {
+                    namedSectionBuilder.Add(getSection(sectionName));
+                }
+                
+                // create the global config
+                GlobalAnalyzerConfig globalConfig = new GlobalAnalyzerConfig(globalSection, namedSectionBuilder.ToImmutableAndFree());
+                _values.Free();
+                return globalConfig;
+
+                Section getSection(string sectionName)
+                {
+                    var dict = _values[sectionName];
+                    var result = dict.ToImmutableDictionary(d => d.Key, d => d.Value.value, Section.PropertiesKeyComparer);
+                    return new Section(sectionName, result);
+                }
+            }
+
+            private void MergeSection(string configPath, Section section)
+            {
+                Debug.Assert(_values is object);
+                Debug.Assert(_duplicates is object);
+
+                if (!_values.TryGetValue(section.Name, out var sectionDict))
+                {
+                    sectionDict = ImmutableDictionary.CreateBuilder<string, (string, string)>(Section.NameEqualityComparer);
+                    _values.Add(section.Name, sectionDict);
+                }
+
+                var duplicateDict = _duplicates.ContainsKey(section.Name) ? _duplicates[section.Name] : null;
```

Comment:
This is a double lookup. Should rewrite to use `TryGetValue`.  That will add an `if` block though so I wouldn't be opposed to a handy extension method here.